### ISSUE TITLE
Initializers of stable constants can call const unstable functions

### ISSUE THREAD

**Issue by ecstatic-morse** [rust-lang/rust#68743]
Crates using `staged_api` are forbidden from calling const unstable functions from stable const functions. However, this restriction [does not extend](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=53d0049a668c92684fd26fc76eafad21) to the initializers of consts or statics. Although unlikely, this could allow for "backdoor stabilization" of various const-eval features. The following example would lock us in to the current implementation of `const_if_match`, for example.

```rust
#![stable(feature = "bar", since = "1.34")]

#![feature(staged_api)]
#![feature(const_if_match)]
#![feature(foo)]

#[rustc_const_unstable(feature = "foo", issue = "0")]
const fn foo() -> i32 {
    if true { 0 } else { 1 }
}

#[stable(feature = "bar", since = "1.34")]
pub const BAR: i32 = foo();
```

cc @rust-lang/wg-const-eval

**Comment by ecstatic-morse** [comment]
I'm mostly wondering if this is intended, and if so, whether we could write down a set of rules to describe the expected behavior.

**Comment by ecstatic-morse** [comment]
It's not just calling unstable functions btw, there's no checks for `rustc_allow_internal_unstable`, so [one could stabilize a const with `&mut`](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=f7d88dcc2b43fd1ce1ca6836ccc9072f).

**Comment by Centril** [comment]
Archaeology time! We sorta decided to not inspect `const`s and `static`s in https://github.com/rust-lang/rust/pull/53604#issuecomment-417619694 when we were first preparing to ship the stable `const fn` fragment. Note in particular (perhaps not true anymore?):

> So... while I could actually implement a check that validates the MIR of constants in `min_const_fn` terms, this would be a major undertaking, requiring a new query and very hacky inspection of the MIR of constants at weird intervals.

I would be all for adding these checks if compiler changes have made those feasible to implement in a clean way.

**Comment by ecstatic-morse** [comment]
Makes sense, it's feasible now. Merging `qualify_min_const_fn` and `check_consts` would make my life easier, but it's kind of a pain to review since a lot of tests will have to change.
